File: python/searchable-analysis/src/searchable_analysis/language_profile.py
import unicodedata
from dataclasses import dataclass
from typing import Callable

from searchable_analysis.segment_cjk import segment_cjk_bigram
from searchable_analysis.segment_latin import segment_latin_words
from searchable_analysis.segment_sea import segment_sea_trigram
from searchable_analysis.stemmer_de import stem_german as _stem_german
from searchable_analysis.stemmer_en import stem_english as _stem_english
from searchable_analysis.stemmer_nl import stem_dutch as _stem_dutch
from searchable_analysis.stemmer_no import stem_norwegian as _stem_norwegian
from searchable_analysis.stemmer_sv import stem_swedish as _stem_swedish
from searchable_analysis.token_span import TokenSpan


@dataclass(frozen=True)
class LanguageProfile:
    code: str
    segment: Callable[[str], list[TokenSpan]]
    fold_diacritics: bool
    stopwords: frozenset
    stem: Callable[[str], str]


def strip_diacritics(term: str) -> str:
    decomposed = unicodedata.normalize("NFKD", term)
    return "".join(c for c in decomposed if not unicodedata.category(c).startswith("M"))


def _identity(term: str) -> str:
    return term


# Standard English function/interrogative words: they carry no topical
# meaning on their own, but under BM25's corpus-relative IDF they can still
# outscore genuinely rare content words in a small, specialized corpus where
# they appear infrequently in formal prose (e.g. "what does X mean" against
# a technical reference). Filtering them at analysis time -- for both
# indexing and querying, since analyze() is shared -- keeps queries focused
# on content words without needing per-corpus tuning. Based on the classic
# Lucene/SMART English stopword list, extended with common
# question/auxiliary words relevant to natural-language questions.
_ENGLISH_STOPWORDS = frozenset(
    {
        "a", "an", "and", "are", "as", "at", "be", "but", "by", "can",
        "could", "did", "do", "does", "for", "he", "how", "i", "if", "in",
        "into", "is", "it", "mean", "meaning", "me", "no", "not", "of",
        "on", "or", "please", "she", "should", "such", "tell", "that",
        "the", "their", "then", "there", "these", "they", "this", "to",
        "was", "we", "what", "when", "where", "which", "who", "whom",
        "will", "with", "would", "you",
    }
)

english = LanguageProfile(
    code="en",
    segment=segment_latin_words,
    fold_diacritics=False,
    stopwords=_ENGLISH_STOPWORDS,
    stem=_stem_english,
)

german = LanguageProfile(
    code="de",
    segment=segment_latin_words,
    fold_diacritics=False,
    stopwords=frozenset(),
    stem=_stem_german,
)

swedish = LanguageProfile("sv", segment_latin_words, False, frozenset(), _stem_swedish)
dutch = LanguageProfile("nl", segment_latin_words, False, frozenset(), _stem_dutch)
norwegian_bokmal = LanguageProfile("nb", segment_latin_words, False, frozenset(), _stem_norwegian)
norwegian_nynorsk = LanguageProfile("nn", segment_latin_words, False, frozenset(), _stem_norwegian)
norwegian = LanguageProfile("no", segment_latin_words, False, frozenset(), _stem_norwegian)

chinese = LanguageProfile(
    code="zh",
    segment=segment_cjk_bigram,
    fold_diacritics=False,
    stopwords=frozenset(),
    stem=_identity,
)

japanese = LanguageProfile(
    code="ja",
    segment=segment_cjk_bigram,
    fold_diacritics=False,
    stopwords=frozenset(),
    stem=_identity,
)

thai = LanguageProfile(
    code="th",
    segment=segment_sea_trigram,
    fold_diacritics=False,
    stopwords=frozenset(),
    stem=_identity,
)

khmer = LanguageProfile(
    code="km",
    segment=segment_sea_trigram,
    fold_diacritics=False,
    stopwords=frozenset(),
    stem=_identity,
)

lao = LanguageProfile(
    code="lo",
    segment=segment_sea_trigram,
    fold_diacritics=False,
    stopwords=frozenset(),
    stem=_identity,
)

File: packages/searchable/src/analysis/language-profile.ts
import { segmentCjkBigram } from "./segment-cjk.js";
import { segmentSeaTrigram } from "./segment-sea.js";
import { stemGerman } from "./stemmer-de.js";
import { stemEnglish } from "./stemmer-en.js";
import { stemDutch } from "./stemmer-nl.js";
import { stemNorwegian } from "./stemmer-no.js";
import { stemSwedish } from "./stemmer-sv.js";
import {
  CJK_STOPWORDS,
  DUTCH_STOPWORDS,
  ENGLISH_STOPWORDS,
  GERMAN_STOPWORDS,
  NORWEGIAN_STOPWORDS,
  SWEDISH_STOPWORDS,
  THAI_STOPWORDS,
} from "./stopwords.js";

export interface TokenSpan {
  text: string;
  isWordLike: boolean;
}

/**
 * Per-language analysis behavior. Indexer and runtime both consume the
 * same profile via `analyze()` so index-time and query-time tokenization
 * can never drift apart (see docs/guides/internationalization.md).
 */
export interface LanguageProfile {
  code: string;
  segment(text: string): TokenSpan[];
  foldDiacritics: boolean;
  stopwords: ReadonlySet<string>;
  stem(term: string): string;
}

function segmentWithIntl(code: string): (text: string) => TokenSpan[] {
  const segmenter = new Intl.Segmenter(code, { granularity: "word" });
  return (text: string) =>
    Array.from(segmenter.segment(text), (s) => ({
      text: s.segment,
      isWordLike: s.isWordLike ?? false,
    }));
}

// Unicode "Mark" category: combining diacritics left behind by NFKD
// decomposition (e.g. \u00e9 -> e + combining acute accent).
const COMBINING_MARKS = /\p{M}/gu;

function stripDiacritics(term: string): string {
  return term.normalize("NFKD").replace(COMBINING_MARKS, "");
}

/**
 * Real affix-stripping stemming (docs/guides/internationalization.md#stemming)
 * via the classic Porter algorithm (`./stemmer-en.ts`) — filtering standard
 * function/interrogative words at analysis time (shared by indexer and runtime).
 */
export const english: LanguageProfile = {
  code: "en",
  segment: segmentWithIntl("en"),
  foldDiacritics: false,
  stopwords: ENGLISH_STOPWORDS,
  stem: stemEnglish,
};

/**
 * Real affix-stripping stemming via the Snowball German algorithm
 * (`./stemmer-de.ts`) — filtering standard function/interrogative words
 * at analysis time. `foldDiacritics: false` below is a *separate*
 * concern that no longer fully delivers on its original promise now
 * that stemming is real: that flag only controls whether raw,
 * *unstemmed* text folds diacritics before analysis; it says nothing
 * about what the stemmer itself does afterward. The Snowball
 * algorithm's own final step unconditionally folds any remaining
 * ä/ö/ü to a/o/u — standard, spec-correct behavior (matches Lucene's
 * and PyStemmer's German stemmers too) — so `schon` and `schön` now
 * both stem to `"schon"` even though `foldDiacritics: false` keeps
 * them distinct going *into* `stem()`. See
 * docs/guides/internationalization.md#case-folding-and-diacritics for the full
 * writeup of this tradeoff.
 */
export const german: LanguageProfile = {
  code: "de",
  segment: segmentWithIntl("de"),
  foldDiacritics: false,
  stopwords: GERMAN_STOPWORDS,
  stem: stemGerman,
};

export const swedish: LanguageProfile = {
  code: "sv",
  segment: segmentWithIntl("sv"),
  foldDiacritics: false,
  stopwords: SWEDISH_STOPWORDS,
  stem: stemSwedish,
};

export const dutch: LanguageProfile = {
  code: "nl",
  segment: segmentWithIntl("nl"),
  foldDiacritics: false,
  stopwords: DUTCH_STOPWORDS,
  stem: stemDutch,
};

export const norwegianBokmal: LanguageProfile = {
  code: "nb",
  segment: segmentWithIntl("nb"),
  foldDiacritics: false,
  stopwords: NORWEGIAN_STOPWORDS,
  stem: stemNorwegian,
};

export const norwegianNynorsk: LanguageProfile = {
  code: "nn",
  segment: segmentWithIntl("nn"),
  foldDiacritics: false,
  stopwords: NORWEGIAN_STOPWORDS,
  stem: stemNorwegian,
};

/** Generic Norwegian compatibility code. Detection returns `nb` or `nn`. */
export const norwegian: LanguageProfile = {
  code: "no",
  segment: segmentWithIntl("no"),
  foldDiacritics: false,
  stopwords: NORWEGIAN_STOPWORDS,
  stem: stemNorwegian,
};

/**
 * Bigram (n-gram) fallback segmentation (`./segment-cjk.ts`) rather
 * than dictionary-based `Intl.Segmenter("zh"|"ja")` word segmentation
 * (docs/guides/internationalization.md#segmentation) — guarantees correct
 * substring matching in any environment without a bundled dictionary,
 * at the cost of index size (overlapping bigrams) and some relevance
 * precision (no real word boundaries), a documented, accepted
 * tradeoff. `stem` is the identity function: Chinese and Japanese
 * don't have Indo-European-style inflectional morphology that
 * affix-stripping stemming targets, matching every other
 * non-affix-stripping language this project documents as a stemming
 * no-op. `foldDiacritics` is meaningless for these scripts (no
 * diacritics), left `false` for consistency with every profile that
 * isn't specifically opting in.
 */
export const chinese: LanguageProfile = {
  code: "zh",
  segment: segmentCjkBigram,
  foldDiacritics: false,
  stopwords: CJK_STOPWORDS,
  stem: (term) => term,
};

/** Same bigram-fallback segmentation as `chinese` above -- the fallback mechanism doesn't depend on which CJK language the text is in, only its script (docs/guides/internationalization.md#segmentation). */
export const japanese: LanguageProfile = {
  code: "ja",
  segment: segmentCjkBigram,
  foldDiacritics: false,
  stopwords: CJK_STOPWORDS,
  stem: (term) => term,
};

/**
 * Trigram (n-gram) fallback segmentation (`./segment-sea.ts`), the same
 * robustness-net strategy as `chinese`/`japanese` above but at width 3
 * rather than 2 (see that module's own doc comment for why). `stem` is
 * the identity function for the same reason as `chinese`/`japanese`:
 * none of these scripts have Indo-European-style inflectional
 * morphology that affix-stripping stemming targets. `foldDiacritics` is
 * deliberately `false`: Thai/Khmer/Lao vowel signs and tone marks are
 * Unicode combining marks but are *not* decorative diacritics -- they
 * distinguish otherwise-identical words, so running them through
 * `stripDiacritics()`'s NFKD-then-strip-combining-marks logic (built for
 * Latin-script accents) would silently conflate distinct words.
 */
export const thai: LanguageProfile = {
  code: "th",
  segment: segmentSeaTrigram,
  foldDiacritics: false,
  stopwords: THAI_STOPWORDS,
  stem: (term) => term,
};

/** Same trigram-fallback segmentation as `thai` above -- the fallback mechanism doesn't depend on which of the three scripts the text is in (docs/guides/internationalization.md#segmentation). */
export const khmer: LanguageProfile = {
  code: "km",
  segment: segmentSeaTrigram,
  foldDiacritics: false,
  stopwords: CJK_STOPWORDS,
  stem: (term) => term,
};

/** Same trigram-fallback segmentation as `thai` above -- the fallback mechanism doesn't depend on which of the three scripts the text is in (docs/guides/internationalization.md#segmentation). */
export const lao: LanguageProfile = {
  code: "lo",
  segment: segmentSeaTrigram,
  foldDiacritics: false,
  stopwords: CJK_STOPWORDS,
  stem: (term) => term,
};

export { stripDiacritics };
